crowbar_framework/lib/crowbar/backup/export.rb

```diff
@@ -0,0 +1,192 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef"
+require "yaml_db"
+
+module Crowbar
+  class Backup
+    class Export < Base
+      attr_accessor :path
+
+      def initialize(path)
+        self.path = path
+      end
+
+      def export
+        clients
+        nodes
+        roles
+        databags
+        db
+        crowbar
+        meta
+      end
+
+      def clients
+        chef(
+          "client",
+          "::Chef::ApiClient"
+        )
+      end
+
+      def nodes
+        chef(
+          "node",
+          "::Chef::Node"
+        )
+      end
+
+      def roles
+        chef(
+          "role",
+          "::Chef::Role"
+        )
+      end
+
+      def databags
+        logger.debug "Backing up databags"
+
+        data_dir = workdir.join("knife", "databags")
+        data_dir.mkpath unless data_dir.directory?
+
+        Chef::DataBag.list.each do |name, url|
+          bag_dir = data_dir.join(name)
+          bag_dir.mkpath unless bag_dir.directory?
+
+          Chef::DataBag.load(name).each do |item, item_url|
+            logger.debug "Backing up databag #{name}/#{item}"
+
+            record = Chef::DataBagItem.load(
+              name,
+              item
+            )
+
+            bag_dir.join("#{item}.json").open("w") do |file|
+              file.write JSON.pretty_generate(
+                record
+              )
+            end
+          end
+        end
+      end
+
+      def db
+        logger.debug "Backing up database"
+
+        data_dir = workdir.join("crowbar")
+        data_dir.mkpath unless data_dir.directory?
+
+        SerializationHelper::Base.new(
+          YamlDb::Helper
+        ).dump(
+          data_dir.join(
+            "database.#{YamlDb::Helper.extension}"
+          )
+        )
+      end
+
+      def crowbar
+        logger.debug "Backing up Crowbar Files"
+
+        data_dir = workdir.join("crowbar")
+        ["keys", "configs", "root"].each do |folder|
+          absolute_path = data_dir.join(folder)
+          absolute_path.mkpath unless absolute_path.directory?
+        end
+
+        self.class.export_files.each do |filemap|
+          source, destination = filemap
+          if source =~ /resolv.conf/
+            data_dir.join(destination).open("w") do |file|
+              forwarders.each do |forwarder|
+                file.write("nameserver #{forwarder}")
+              end
+            end
+          else
+            # copy files with higher permissions
+            dest = data_dir.join(destination).to_s
+            system("sudo", "-i", "cp", "-r", source, dest)
+            system("sudo", "-i", "chown", "-R", "crowbar:crowbar", source, dest)
+          end
+        end
+      end
+
+      def meta
+        meta = Hash.new
+        meta["version"] = ENV["CROWBAR_VERSION"]
+        meta["created_at"] = Time.zone.now.to_s
```

Comment:
I would probably add the platform of the admin server to the metadata (since in theory, the files might not be located in the same location).